rtos/TARGET_CORTEX/mbed_rtx_idle.cpp

```diff
@@ -97,8 +91,7 @@ static void default_idle_hook(void)
 {
     uint32_t elapsed_ticks = 0;
 
-    core_util_critical_section_enter();
```

Comment:
@c1728p9 we need to review this critical section. RTX5.3 marks the svcRtxKernelSuspend/Resume as static so we can't use it here and I'd rather not modify RTX again. Can you remember why are we entering the critical section before suspending the kernel?

Reply:
The general problem is that the check that it is safe to enter sleep and the act of entering sleep need to be done atomically. If this is not done an interrupt could change state by waking a thread before WFI or WFE has been called.

This problem is visible this [Tick-less Low-Power Operation](https://arm-software.github.io/CMSIS_5/RTOS2/html/theory_of_operation.html) code:
```c
void osRtxIdleThread (void) {
    for (;;) {
        tc_wakeup = osKernelSuspend();
        /* Is there some time to sleep? */                                            
        if (tc_wakeup > 0) {
        tc = 0;
        /* Enter the low power state */
        MSP432_LP_Entry();
                                                //<- If an interrupt signals a thread to wake up at this point will the call to WFE() won't wake because of it. There is a thread that should be running now but won’t be able to until tc_wakeup ticks have passed.
        __WFE();
        }
        /* Adjust the kernel ticks with the amount of ticks slept */
        osKernelResume (tc);
    }
}
```

Reply:
@JonatanAntoni I don't think it's currently possible to protect against this race using RTX APIs, do you guys have a way of dealing with it?

Reply:
__WFE should not have a race here, only __WFI would be problematic to my knowledge.
To my understanding each interrupt that is enabled to set the event register should do so. __WFE returns immediately if the event register is already set. Otherwise it puts the device to sleep until the event flag is set by any active event source. __WFE clears the event register on return.

Reply:
That sounds good to me, I had a look at the Architecture Reference and couple of manuals for specific designs from partners and it seems that no one does anything weird. Anyone against replacing `WFI` with `WFE`? @c1728p9 @0xc0170

Reply:
That solution sounds good to me as well as long as we don't have any devices with the errata mentioned [here](http://www.keil.com/support/docs/3591.htm). I was under the impression that __WFE required any interrupt that could wake the system to call __SEV, which is not the case. This should be set by hardware automatically and its just errata that caused it not to be set for some devices.

Reply:
Lets change it then and resolve this internal API usage